File: Search/DFS/combination-sum/combination-sum.js
function combinationSum(candidates, target) {
  if (!candidates) {
      return [];
  }
  let copyedCandidates = candidates.sort();

  let results = [];

  dfs(results, [], 0, target, copyedCandidates);

  return results;
}

function dfs(results, local, startIndex, target, candidates) {
  if (target === 0) {
    results.push(local.slice());
    return;
  }

  if (target < 0) {
    return 
  }

  for (let i = startIndex; i < candidates.length; i++) {
    local.push(candidates[i]);
    dfs(results, local, i, target - candidates[i], candidates);
    local.pop();
  }

}


const expectedResult = [
  [2,2,3],
  [7]
]
console.log(combinationSum([2,3,6,7], 7));
console.log(combinationSum([2, 3, 5], 8));

File: algorithms/search/Search/DFS/combination-sum/combination-sum.py
class Solution:
  
  def combinationSum(self, candidates, target):
    if not candidates:
      return []
    
    candidates.sort()
    result = []
    self.dfs(candidates, 0, target, [], result)


  def dfs(self, candidates, startIndex, target, localCombos, result):
    if target == 0:
      result.append(localCombos[:])
      return
    
    for i in range(startIndex, len(candidates)):
      if i > startIndex and candidates == candidates[i - 1]:
        continue
      
      if candidates[i] > target:
        return

    localCombos.append(candidates[i])
    self.dfs(candidates, i, target - candidates[i], localCombos, result)
    localCombos.pop()